### PR TITLE
Fix level refactor

### PR DESCRIPTION
**Pull request by chanoiano**

### CODE REVIEW

File: src/main/java/core/levels/LevelManager.java [hunk 1]
```diff
@@ -110,12 +118,50 @@ private void updateEnemies() {
...
+        if (level.getCurrentTime() == 0) {
+            for (int i = 0; i < numOfEnemies; i++) {
+                Spawner<Character> enemySpawner = createEnemySpawner();
+                enemySpawner.addSpawnCallback(character -> character.setTeam("enemy"));
+                enemies.add(enemySpawner);
+            }
+        } else{
+            double numEnemies = Math.floor(level.getCurrentTime() / level.getSpawnTime());
+            for (int i = 0; i < numEnemies; i++) {
+                enemies = enemySpawns;
```
Comment (mdawess) [inline src/main/java/core/levels/LevelManager.java]: Why is update enemies not used here?

File: src/main/java/core/levels/LevelState.java [hunk 2]
```diff
@@ -89,11 +88,8 @@ public void finishedLevel() {
         levelFinished = true;
     }
 
-    public List<Spawner<Character>> getEnemySpawns() {
-        return enemySpawns;
-    }
+    public String getLevelDifficulty() { return levelDifficulty; }
```
Comment (mdawess) [inline src/main/java/core/levels/LevelState.java]: Much more in line with CLEAN, great job!

File: src/main/java/core/levels/LevelManager.java [hunk 3]
```diff
@@ -37,12 +44,13 @@ public LevelManager(LevelState level) {
 
         this.batch = new SpriteBatch();
 
+        this.levelDifficultyToEnemies = createLevelDifficultyToEnemies();
+        // create enemy list based on difficulty of level
+        this.enemySpawns = createEnemyList(levelDifficultyToEnemies.get(level.getLevelDifficulty()));
```
Comment (matrix-marker) [inline src/main/java/core/levels/LevelManager.java]: this needs to be more abstract – should be able to spawn more than just enemies, and name should reflect this

File: src/main/java/core/levels/LevelManager.java [hunk 4]
```diff
@@ -110,12 +118,50 @@ private void updateEnemies() {
...
         }
     }
 
+    /**
+     * Create HashMap holding levelDifficulty as key, and number of enemies that correspond to
+     * that levelDifficulty, as value
+     *
+     * @return HashMap containing levelDifficulty mapped to number of enemies
+     */
+    private static HashMap<String, Integer> createLevelDifficultyToEnemies() {
```
Comment (matrix-marker) [inline src/main/java/core/levels/LevelManager.java]: unclear method name

File: src/main/java/core/levels/LevelManager.java [hunk 5]
```diff
@@ -110,12 +118,50 @@ private void updateEnemies() {
...
             enemy.spawn();
-            level.setEnemySpawns(enemies);
+            enemySpawns = enemies;
             level.setScore(level.getScore() + 1);
             level.setSpawnTime(level.getSpawnTime() + 15);
         }
     }
 
+    /**
+     * Create HashMap holding levelDifficulty as key, and number of enemies that correspond to
```
Comment (matrix-marker) [inline src/main/java/core/levels/LevelManager.java]: levelDifficulty being a key makes no sense to me

why is it called levelDifficulty? then also levels with same difficulty would have the same key? also this is a refactor, adding LevelDifficulty seems to imply a whole new feature

just unclear what's it's used for

File: src/main/java/core/levels/LevelManager.java [hunk 6]
```diff
@@ -110,12 +118,50 @@ private void updateEnemies() {
...
+        }
+        return my_dict;
+    }
+
+    /**
+     * Creates list of enemies to be spawned in level
+     * @param numOfEnemies wanted to be spawned in this level
+     * @return enemies list
+     */
+    private List<Spawner<Character>> createEnemyList(int numOfEnemies) {
```
Comment (matrix-marker) [inline src/main/java/core/levels/LevelManager.java]: should be createSpawnList or smth, that reflects creating spawns of any kind of WorldEntity

File: src/main/java/core/levels/LevelManager.java [hunk 7]
```diff
@@ -97,7 +105,7 @@ public void step(float dt) {
      */
     private void updateEnemies() {
```
Comment (matrix-marker) [inline src/main/java/core/levels/LevelManager.java:151]: unclear method name

File: src/main/java/core/levels/LevelManager.java [hunk 8]
```diff
@@ -110,12 +118,50 @@ private void updateEnemies() {
         if (level.getCurrentTime() >= level.getSpawnTime()) {
             Spawner<Character> enemy = enemies.remove(0);
             enemy.spawn();
-            level.setEnemySpawns(enemies);
+            enemySpawns = enemies;
             level.setScore(level.getScore() + 1);
```
Comment (matrix-marker) [inline src/main/java/core/levels/LevelManager.java:166]: level score is being mutated in a method called UpdateEnemies, not sure why

Comment (matrix-marker) [reply src/main/java/core/levels/LevelManager.java:166]: should be moved or removed

File: src/main/java/core/levels/LevelManager.java [hunk 9]
```diff
@@ -110,12 +118,50 @@ private void updateEnemies() {
...
             level.setSpawnTime(level.getSpawnTime() + 15);
```
Comment (matrix-marker) [inline src/main/java/core/levels/LevelManager.java]: unclear what this does

File: src/main/java/core/levels/LevelManager.java [hunk 10]
```diff
@@ -110,12 +118,50 @@ private void updateEnemies() {
...
     }
 
+    /**
+     * Create HashMap holding levelDifficulty as key, and number of enemies that correspond to
+     * that levelDifficulty, as value
+     *
+     * @return HashMap containing levelDifficulty mapped to number of enemies
+     */
+    private static HashMap<String, Integer> createLevelDifficultyToEnemies() {
+        HashMap<String, Integer> my_dict = new HashMap<>();
```
Comment (matrix-marker) [inline src/main/java/core/levels/LevelManager.java]: uninformative variable name, maybe use `difficultyDict` or similar

File: src/main/java/core/levels/LevelManager.java [hunk 11]
```diff
@@ -110,12 +118,50 @@ private void updateEnemies() {
...
+    }
+
+    /**
+     * Creates list of enemies to be spawned in level
+     * @param numOfEnemies wanted to be spawned in this level
+     * @return enemies list
+     */
+    private List<Spawner<Character>> createEnemyList(int numOfEnemies) {
+        List<Spawner<Character>> enemies = new ArrayList<>();
+        if (level.getCurrentTime() == 0) {
```
Comment (matrix-marker) [inline src/main/java/core/levels/LevelManager.java]: createEnemyList depending on level time is weird imo: we're creating enemy spawners anyways, so the behaviour shouldnt be that different

File: src/main/java/core/levels/LevelState.java [hunk 12]
```diff
@@ -18,8 +17,8 @@
     private boolean levelFinished;
     protected transient float currentTime;
     protected int score;
-    protected List<Spawner<Character>> enemySpawns;
     private float spawnTime;
```
Comment (matrix-marker) [inline src/main/java/core/levels/LevelState.java]: rename to spawnInterval, add comment specifying the unit is seconds

**Review by matrix-marker — CHANGES_REQUESTED** [review]

File: src/main/java/core/levels/LevelManager.java (hunk 1 above)
Comment (chanoiano) [reply src/main/java/core/levels/LevelManager.java]: updateEnemies is for spawning the enemy into the world itself, whereas createEnemyList is just to populate the list from which updateEnemies takes enemies to spawn. will make clearer comments for documentation to reflect this

**Pull request merged**